File: frontend/src/api/knowledge.ts
import request from './request'
import { BaseAPI } from './base'

export interface KnowledgeDoc {
  id?: number
  project_id: number
  title: string
  content?: string
  file_type?: string
  file_path?: string
  file_size?: number
  source_type?: string
  source_id?: number
  chunk_count?: number
  chunk_strategy?: string
  chunk_size?: number
  overlap?: number
  status?: string
  error_message?: string
  created_by?: number
  created_at?: string
  updated_at?: string
}

export interface KnowledgeChunk {
  id: number
  doc_id: number
  project_id?: number
  chunk_index: number
  content: string
  token_count: number
  doc_title: string
  created_at?: string
}

export interface KnowledgeChunkListResponse {
  total: number
  page: number
  page_size: number
  items: KnowledgeChunk[]
}

export interface KnowledgeDocListResponse {
  total: number
  page: number
  page_size: number
  items: KnowledgeDoc[]
}

export interface KnowledgeSearchResult {
  query: string
  results: Array<{
    doc_id: number
    title: string
    content: string
    chunk_index: number
    score: number
    similarity: number
  }>
  total: number
}

export interface KnowledgeStats {
  project_id: number
  total_docs: number
  total_chunks: number
}

/** 知识库文档标准 CRUD（list 因路径 /docs/search 非标准，保留自定义） */
const knowledgeApi = new BaseAPI<KnowledgeDoc>('/knowledge')

export function getKnowledgeDocs(projectId: number, params?: { page?: number; page_size?: number; keyword?: string; source_type?: string }) {
  return knowledgeApi.post<KnowledgeDocListResponse>(`/projects/${projectId}/knowledge/docs/search`, params)
}

export function getKnowledgeChunks(projectId: number, params: { page?: number; page_size?: number; keyword?: string; doc_id?: number }) {
  return request.post<KnowledgeChunkListResponse>(`/projects/${projectId}/knowledge/chunks/search`, params)
}

export function syncRequirementsToKnowledge(projectId: number, requirementIds?: number[]) {
  return request.post<{ synced: number; doc_ids: number[]; message: string }>(
    `/projects/${projectId}/knowledge/sync-requirements`,
    requirementIds?.length ? requirementIds : [],
  )
}

export function getKnowledgeDoc(projectId: number, docId: number) {
  return knowledgeApi.get(projectId, docId)
}

export function createKnowledgeDoc(projectId: number, data: { title: string; content: string; file_type?: string }) {
  return knowledgeApi.create(projectId, data)
}

export function uploadKnowledgeDoc(projectId: number, file: File) {
  const formData = new FormData()
  formData.append('file', file)
  return request.post<KnowledgeDoc>(`/projects/${projectId}/knowledge/upload`, formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
}

export function deleteKnowledgeDoc(projectId: number, docId: number) {
  return knowledgeApi.remove(projectId, docId)
}

export function generateChunks(projectId: number, docId: number) {
  return request.post<{ doc_id: number; status: string; message: string }>(
    `/projects/${projectId}/knowledge/${docId}/generate-chunks`,
  )
}

export function searchKnowledge(projectId: number, query: string, topK = 5) {
  return request.post<KnowledgeSearchResult>(`/projects/${projectId}/knowledge/search`, { query, top_k: topK })
}

export function getKnowledgeStats(projectId: number) {
  return request.get<KnowledgeStats>(`/projects/${projectId}/knowledge/stats`)
}

File: aits_hub_python/app/schemas/knowledge.py
"""
知识库 Pydantic 模型
"""
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field


class KnowledgeDocBase(BaseModel):
    title: str = Field(..., max_length=200, description="文档标题")
    content: str = Field(default="", description="文档内容")
    file_type: str = Field(default="text", description="文件类型")


class KnowledgeDocCreate(KnowledgeDocBase):
    pass


class KnowledgeDocResponse(KnowledgeDocBase):
    id: int
    project_id: int
    file_path: str = ""
    file_size: int = 0
    source_type: str = "manual"
    source_id: Optional[int] = None
    chunk_count: int = 0
    chunk_strategy: str = "fixed"
    chunk_size: int = 500
    overlap: int = 50
    status: str
    error_message: str = ""
    created_by: Optional[int] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class KnowledgeDocListResponse(BaseModel):
    total: int
    page: int
    page_size: int
    items: List[KnowledgeDocResponse]


class KnowledgeChunkResponse(BaseModel):
    """切片响应"""
    id: int
    doc_id: int
    project_id: int = 0
    chunk_index: int
    content: str
    token_count: int = 0
    doc_title: str = ""
    created_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class KnowledgeChunkListResponse(BaseModel):
    """切片列表响应"""
    total: int
    page: int
    page_size: int
    items: List[KnowledgeChunkResponse]


class KnowledgeSearchRequest(BaseModel):
    query: str
    top_k: int = Field(default=5, ge=1, le=20)


class KnowledgeSearchResult(BaseModel):
    """单条检索结果"""
    doc_id: int
    chunk_id: int
    title: str = ""
    content: str
    chunk_index: int = 0
    score: float = 0.0
    similarity: float = 0.0


class KnowledgeSearchResponse(BaseModel):
    query: str
    results: List[Dict[str, Any]]
    total: int


class KnowledgeStatsResponse(BaseModel):
    project_id: int
    total_docs: int
    total_chunks: int
